AliTPCCommon

rsadikin opened the pull request:
Dear Ernst

Here I implement irregular grid for correction, and make some debugging on local distortions as you suggested.

Regards
Rifki

Commit:
Distortions -- irregular interpolation fix with kdtree

Add kdtree to irregular interpolator

Fix correction calculation of RPhi

Distortions modification as Ernst suggest

Integration on exact function

davidrohr committed:
Major refactoring of Tracker Framework code, merge nested old framework classes into AliGPUReconstruction, fix plenty of indentation

ehellbar commented:
Hello Rifki,

can you do a git push -f, please, to update the list of changes? My master branch was not up to date so there are also a lot of older commits in this PR. 

Cheers,
Ernst

davidrohr committed:
Fix memory allocation, take into account alignment for required memory size

rsadikin commented:
Hi Ersnt,

I have no permission to write to your master branch.

remote: Permission to ehellbar/AliTPCCommon.git denied to rsadikin.

Or is it something else?
Regards
Rifki

sgorbuno committed:
bug fix in initialization of distortion flag + small fixes

davidrohr committed:
BugFix: Should allocate constant memory as uint4 instead of char to ensure correct alignment

davidrohr committed:
Temporary fix for OpenCL: crashes when compiled in debug mode - disalbe -g

davidrohr committed:
Add flag to control global init mutex

ehellbar commented on TPCSpaceChargeBase/AliTPCSpaceCharge3DCalc.cxx:
```diff
@@ -0,0 +1,4178 @@
+/// First, we calculate integration of the Electric field in z-direction for all direction.
+/// Assumption: \f$ z_{0} \f$ is location of CE (Central Electrode) and \f$ z_{nZColumn - 1} \f$ is location of End Plate.
+///
+/// This integration is in \f$z\f$ direction we can only 
```
According to the convention, this should be
`-1 * dPhi * (radius0 + drDist)`
because the radius of the irregular point is radius0 + drDist.

rsadikin replied:
> According to the convention, this should be
> `-1 * dPhi * (radius0 + drDist)`
> because the radius of the irregular point is radius0 + drDist.

Hello Ernst, sorry I don't understand, so I need to change the line in 1498-1501 to your comment?

ehellbar replied:
Hello Rifki, 
for some reason, the wrong lines are shown in the comment section, they do not correspond to the lines of my comments.

If you click on "Files changed" button on the top of the page and then look into AliTPCSpaceCharge3DCalc.cxx, the comments will be below the correct lines.

ehellbar commented on TPCSpaceChargeBase/AliTPCSpaceCharge3DCalc.cxx:
```diff
@@ -0,0 +1,4178 @@
+/// First, we calculate integration of the Electric field in z-direction for all direction.
+/// Assumption: \f$ z_{0} \f$ is location of CE (Central Electrode) and \f$ z_{nZColumn - 1} \f$ is location of End Plate.
+///
+/// This integration is in \f$z\f$ direction we can only 
```
Can you remove the comment?

ehellbar replied:
Can you remove the comment, please?

ehellbar replied:
Also here according to convention: `-dPhi * (radius0 + drDist)`

ehellbar replied:
This is the regular lookup at radius i which is radius0, so:
`dPhi = (*mCorrDPhiRDz)(i, j + 1) / radius0;`

ehellbar replied:
First calculate the new phi position because it requires the original radius to convert drphi to dphi
`    pCyl[1] = pCyl[1] + fCorrectionFactor * dCyl[1] / pCyl[0];`
`    pCyl[0] = pCyl[0] + fCorrectionFactor * dCyl[0];`

ehellbar replied:
Same as for GetDistortion. First calculate new phi position, then new radial position.

davidrohr committed:
Add missing GPU error handling

davidrohr committed:
Remove Vc defines from SliceData, which were never used

davidrohr committed:
Make option settable to require GPU device for OpenCL

davidrohr committed:
Add setting for OpenCL platform ID

davidrohr committed:
Make Standalone Display compatible to AliGPUReconstruction, encapsulate display in classes

davidrohr committed:
Make Standalone QA compatible to AliGPUReconstruction

davidrohr committed:
Add .gitignore

davidrohr committed:
Improve documentation, fix uninitialized variable

davidrohr committed:
Get rid of old main include header

davidrohr committed:
Some build fixes, add option to build event display / QA with O2

davidrohr committed:
Make QA and Display configurable through config ptr in AliGPUReconstruction, extract configuration structure from standalone qconfigoptions.h file

davidrohr committed:
Bugfix: Install missing cmodules headers in correct subfolder

davidrohr committed:
Workaround for false GCC warnings for sprintf arguments in variadic template

davidrohr committed:
Keep AliGPUReconstruction ptr in Merger also for CPU processing

davidrohr committed:
Bugfix: add missing dummy functions if O2 headers are unavailable

davidrohr committed:
fix memory leak

davidrohr committed:
Expose all configuration headers and let the O2 CA Tracking interface be configured by passing the configuration objects instead of the stupid config string

rsadikin committed:
Add correction with side information

davidrohr committed:
Set some defines here so we do not need to hack them into O2

davidrohr committed:
Remove all HLTCA macros to GPUCA, since this is not HLT any more

davidrohr committed:
Add missing files I forgot to add

davidrohr committed:
remove deleted files from linkdef

davidrohr committed:
Cleanup of macros and headers

 - GPUCA_O2_LIB does no longer imply GPUCA_STANDALONE, as these are two different build
 - Several profiling/timing headers are no longer protectected by GPUCA_STANDALONE #ifdefs.
 - Instead, they are compiled normally but unused.
 - Get rid of weird ../FOLDER/FILE structure for some includes, but add base folder as include directory

davidrohr committed:
Fix compilation without GPU tracker

davidrohr committed:
Simplify / remove unneeded functions from AliGPUReconstruction

davidrohr committed:
Remove functionality to dump timeframe as O2 clusters, no longer needed

davidrohr committed:
Restore outputControl memory functionality to standalone benchmark after switch from AliHLTTPCCAStandaloneFramework to AliGPUReconstruction

davidrohr committed:
Use unique_ptr in standalone benchmark

davidrohr committed:
Fix macro redifinition error in O2

davidrohr committed:
Use snprintf to avoid memory overflow where external string input is possible

davidrohr committed:
fix missing clusterNative in clear / allocate io ptrs

davidrohr committed:
Fix compilation without ROOT / QA / Event Display / O2

davidrohr committed:
Refacture code for timeframe overlay and main benchmark file

davidrohr committed:
fix correct destruction order of CUDA objects (3 independent fixes...)

davidrohr committed:
Adapt Toy MC Generator to new event dump format

davidrohr committed:
ROOT5 does not support scoped enums

davidrohr committed:
Mark virtual functions as override

martenole committed:
Rename TRD tracking components AliHLTTRD* -> AliGPUTRD*

davidrohr committed:
Rename AliHLT to AliGPU

davidrohr committed:
Fix standalone GPU build after renaming AliHLT --> AliGPU

davidrohr committed:
Add AliGPUProcessor class to encapsulate and unify all SetGPUTracker(...) etc

davidrohr committed:
Move merger component to merger folder

davidrohr committed:
Improve AliGPU memory allocation scheme

davidrohr committed:
clean up and unify getPointer function with alignment

davidrohr committed:
Remove unneeded AliHLTArray class and some unused variables

davidrohr committed:
Convert some compile-time define setting to normal settings

davidrohr committed:
swap out dump functionality to AliGPUTPCTrackerDump.cxx, remove some dummy files

davidrohr committed:
Clean up processor initialization, and avoid double initializations

davidrohr committed:
Add category for permanent memory which is not resized, allocate slice data row memory in the same way as the others (now that it is possible)

davidrohr committed:
Use new memory allocation scheme throughout TPCGPUSliceTracker

davidrohr committed:
Save some memory, no need to allocate links memory twice for device/host

davidrohr committed:
With the new memory management, we can enable the standalone event display also for GPU tracking

davidrohr committed:
Register AliGPUProcessors with AliGPUReconstruction and automatize memory registration

davidrohr committed:
TPC GM Merger should use mParam of AliGPUProcessor instead of keeping its own copy

davidrohr committed:
Make memory type a bit-mask for more flexibility

davidrohr committed:
Transition TPC GPU Merger to new memory allocation scheme

davidrohr committed:
Remove remainders from old GPU memory allocation no longer needed

davidrohr committed:
Add GLFW backend to standalone display

davidrohr committed:
Fix alignment of memory allocation for individual allocation strategy

davidrohr committed:
Fix check for keypress in Linux in standalone display

davidrohr committed:
Improve GLUT and GLFW implementations

davidrohr committed:
Set correct GLFW hints for OpenGL version, remove some deprecated OpenGL functionality

davidrohr committed:
Backend can now configure whether it wants to receive external keys from terminal or not

davidrohr committed:
Fix colliding key assignments

davidrohr committed:
Add code to read event dumps in old format

davidrohr committed:
Increase default GPU memory allocation size

davidrohr committed:
Add Clear() functions for Tracker and Merger, fix compilation

davidrohr committed:
Add error handling to event display startup

davidrohr committed:
BugFix: Remove double registration

davidrohr committed:
GL extension loader should depend on the backend

martenole committed:
Prepare GPU tracking TRD

davidrohr committed:
Make GLFW backend of Standalone event display runable from O2

rsadikin commented:
HI Ernst I have changed the code accordingly, however we need to make sure that changes are consistent. So, I am running the test now

If it were consistent,  should I make a new request?

ehellbar commented:
Hello Rifki,
thanks for making the changes! Usually it should be enough if you just make a git push --force on your branch. This PR should then be automatically updated.
Cheers,
Ernst

davidrohr committed:
Homogenize OpenCL and CUDA code, in order to merge common parts in the future

davidrohr committed:
Use variadic template to shorten setting of OpenCL kernel arguments

davidrohr committed:
Bugfix: fix logic and avoid segfault

davidrohr committed:
Reset of NTracklets no longer needed, but break standalone display

davidrohr committed:
Fix memory out of bounds errors during OpenCL GPU synchronization

davidrohr committed:
Add missing synchronization before copying debug data to host/GPU with OpenCL

davidrohr committed:
Fix checks for availability of GPU merger

davidrohr committed:
Clean up treatment of incosecutive cluster IDs in AliRoot HLT

davidrohr committed:
Minor fixes to keypress handling in event display

davidrohr committed:
Port remaining features of track merger to GPU

davidrohr committed:
Add option to disable attachment steps during refit

davidrohr committed:
Get rid of management functionality of AliGPUTPCClusterData class

davidrohr committed:
Fix installation of headers

davidrohr committed:
Copy TPCFastTransform and TRDGeometry to GPU if available

davidrohr committed:
Filter installation of headers in utility folder for only .h files

davidrohr committed:
Extract some device specific GPU code into functions as preparation to merge CUDA and OpenCL code

davidrohr committed:
Add timing measurements for TRD tracking + improve timing output

davidrohr committed:
Extract more device-specific code into functions

davidrohr committed:
Fix compilation with AliRoot and with C++<14

davidrohr committed:
Reduce OpenGL requirement to 4.5 and split away functions for KeyPress handling from main display file

davidrohr committed:
Add protection against running ROOT macros without setting build type

davidrohr committed:
Enforce  definitions int O2 CATracking configuration header

davidrohr committed:
Add required definitions to $DEFINITIONS varialbe such that they are picked up by ROOT generate dictionary

davidrohr committed:
Improvements to O2 backend for standalone event display

davidrohr committed:
Some more renaming

davidrohr committed:
Add some template magic to call templated virtual functions for calling kernels

davidrohr committed:
Pass all definitions from CMake to OpenCL device code compiler

davidrohr committed:
Move memory alignment helpers from AliGPUReconstruction to AliGPUProcessor (simplify AliGPUReconstruction and shorten code in Processors)

davidrohr committed:
Extract some common functionality into PrepareEvent() and SetupGPUProcessor() and Clear()

davidrohr committed:
Add CUSTOM_TRANSFER memory type, fix transfer for TRD GPU tracker

davidrohr committed:
Add functionality to allocated unmanaged memory

davidrohr committed:
Remove obsolete kernel wrappers

davidrohr committed:
Fix Compilation warnings wrt old C-headers and NULL instead of nullptr

davidrohr committed:
Add error handling if TRD tracks/tracklets run out of space

davidrohr committed:
Use common class to hold tpc/trd tracker/merger workers for GPU

davidrohr committed:
Adjust CUDA to use new kernel Launch template

davidrohr committed:
Adjust CPU version to use new kernel Launch template

davidrohr committed:
Minor optimization, start hits beyond NROWS - 6 can never become proper tracks

davidrohr committed:
Fix AliRoot compilation, std:max is not constexpr in C++17 (thanks to Sergey)

rsadikin committed:
Remove comments and update on phi r distortions

davidrohr committed:
Make OpenCL version use constant memory for Parameters and Workers

davidrohr committed:
Fix for CPU version, use number of slices instead of start/end for kernel template, clean up signed/unsigned comparisons

davidrohr committed:
remove obsolete pragmas

davidrohr committed:
Use _WIN32 define exclusively and remove WIN32 (in case standalone version ie ever used again on Windows)

davidrohr committed:
Get rid of STATIC_ASSERT macro and template since we have C++11

davidrohr committed:
Clean up headers/includes

davidrohr committed:
Fix: Forward additional call parameters to kernel (although not yet used)

davidrohr committed:
Adjust OpenCL version to use new kernel Launch template

davidrohr committed:
Add static_assert sanity check that our hack with the deviceEvent typedef is correct

davidrohr committed:
Bugfix: we have to set num to 0 for non-multi version, otherwise OpenCL will call the incorrect multi/non-multi kernel version and pass incorrect arguments

davidrohr committed:
Add event treatment to kernel call template, reorder event parameters for event transfer to match kernel call scheme

davidrohr committed:
Throw exception in case of CUDA / OpenCL / OpenGL failure (return 1 goes unnoticed too often...)

davidrohr committed:
Add event handling on CUDA backend (not needed yet, but bring it to same feature level as OpenCL backend)

davidrohr committed:
Remove rowInit kernels and replace with generic memclean kernels

davidrohr committed:
Add virtual function for copying to constant memory in DeviceBase

davidrohr committed:
Differentiate between failure during initialization (return 1) and elsewhere (throw exception)

davidrohr committed:
Remove old variant of GPU sync points which were not used anymore in that way

davidrohr committed:
Add event creation/destruction to CUDA to provide same events as in OpenCL version, unify stream synchronization via events

davidrohr committed:
Replace plenty of 36 by NSLICES

davidrohr committed:
Add template integer to allow for multiple kernels per kernel class, move alternate processing variant of TrackletConstructor to new kernel scheme

davidrohr committed:
Add runtime settings for number of streams, pipelind mode for constructor / selector. Add alternative pipelined constructor strategy to OpenCL version.

davidrohr committed:
Merge main CUDA and OpenCL reconstruction code

davidrohr committed:
remove obsolete time profile macros and simplify code slightly

davidrohr committed:
Remove obsolete on-the-fly patching of CUDA files

davidrohr committed:
Add HIP Backend for TPC GPU Code

davidrohr committed:
Compile fixes for other make settings

davidrohr committed:
Get rid of all double precision floating point constants + several fixes in preparation for OpenCL-C++

rsadikin committed:
Add distortion z, deleted at the last commit
